atomics/include/desul/atomics/Compare_Exchange_GCC.hpp

```diff
@@ -22,29 +22,38 @@ void atomic_thread_fence(MemoryOrder, MemoryScope) {
   __atomic_thread_fence(GCCMemoryOrder<MemoryOrder>::value);
 }
 
+template <typename T, class MemoryOrder, class MemoryScope>
+T atomic_exchange(
+    T* dest, T value, MemoryOrder, MemoryScope) {
+  T return_val;
+  __atomic_exchange(
+     dest, &value, &return_val, GCCMemoryOrder<MemoryOrder>::value);
+  return return_val;
+}
+
 // Failure mode for atomic_compare_exchange_n cannot be RELEASE nor ACQREL so
 // Those two get handled separatly.
 template <typename T, class MemoryOrder, class MemoryScope>
 T atomic_compare_exchange(
     T* dest, T compare, T value, MemoryOrder, MemoryScope) {
-  (void)__atomic_compare_exchange_n(
-      dest, &compare, value, false, GCCMemoryOrder<MemoryOrder>::value, GCCMemoryOrder<MemoryOrder>::value);
+  (void)__atomic_compare_exchange(
+      dest, &compare, &value, false, GCCMemoryOrder<MemoryOrder>::value, GCCMemoryOrder<MemoryOrder>::value);
   return compare;
 }
 
 template <typename T, class MemoryScope>
 T atomic_compare_exchange(
     T* dest, T compare, T value, MemoryOrderRelease, MemoryScope) {
-  (void)__atomic_compare_exchange_n(
-      dest, &compare, value, false, __ATOMIC_RELEASE, __ATOMIC_RELAXED);
+  (void)__atomic_compare_exchange(
+      dest, &compare, &value, false, __ATOMIC_RELEASE, __ATOMIC_RELAXED);
   return compare;
 }
 
 template <typename T, class MemoryScope>
 T atomic_compare_exchange(
     T* dest, T compare, T value, MemoryOrderAcqRel, MemoryScope) {
-  (void)__atomic_compare_exchange_n(
-      dest, &compare, value, false, __ATOMIC_ACQ_REL, __ATOMIC_ACQUIRE);
+  (void)__atomic_compare_exchange(
+      dest, &compare, &value, false, __ATOMIC_ACQ_REL, __ATOMIC_ACQUIRE);
```

Comment:
Why did you decide to use `__atomic_compare_exchange`?  Not that I prefer one built-in function over the other but as far as I can tell the changes are without any effect.  Am I missing something?

Reply:
__atomic_compare_exchange_n doesn't work for arbitrary types. I misread that initially. Only __atomic_compare_exchange works for arbitrary types, i.e. it just works on echanging sizeof(T) using the three provided pointers.